File: scripts/validate_r_cmd_check_output.R
#!/usr/bin/env Rscript

# be aware that CRAN rules and messages can change with R upgrade

# NOTE whitelist to be used by grepl, for each check step
whitelist = list(
  "CRAN incoming feasibility" = c("Maintainer:",
                                  "Checking URLs requires 'libcurl' support in the R build",
                                  "Package has FOSS license, installs .class/.jar but has no 'java' directory",
                                  "Insufficient package version", # when we check older version of package than current available on CRAN
                                  "Days since last update", # when we submit to CRAN recently
                                  "Number of updates in past 6 months"), # when we submit to CRAN too often
  "package dependencies" = c("No repository set, so cyclic dependency check skipped",
                             "Package suggested but not available for checking"),
  "installed package size" = c("installed size is .*Mb", # h2o.jar is installed
                               "sub-directories of 1Mb or more",
                               "java .*Mb"),
  "Rd cross-references" = "Package unavailable to check Rd xrefs" # when linking documentation to optional deps not present at check time
)
check_note <- function(details, whitelist, verbose=TRUE) {
  stopifnot(is.data.frame(details), is.list(whitelist))
  status_details <- details[details$Status=="NOTE", , drop=FALSE]
  if (!nrow(status_details))
    stop("No NOTEs were found in provided details, be sure to escape this call when no NOTEs present")
  lines <- strsplit(status_details$Output, "\n", fixed=TRUE)
  names(lines) <- status_details$Check
  filter_check <- function(check, lines, whitelist) {
    check_lines <- lines[[check]]
    check_lines <- check_lines[nzchar(check_lines)]
    check_whitelist <- whitelist[[check]]
    # no whitelist for this check step, all notes are valid
    if (!length(check_whitelist))
      return(check_lines)
    # check if whitelisted
    wl <- sapply(
      check_lines,
      function(line) any(sapply(
        check_whitelist,
        function (pattern) grepl(pattern, line)
        ))
    )
    check_lines[!wl]
  }
  new_note <- sapply(names(lines), filter_check, lines, whitelist, simplify=FALSE)
  new_note_row <- which(sapply(new_note, length) > 0L)
  new_note <- new_note[new_note_row]
  new_note_flags <- unique(status_details[new_note_row, "Flags"])
  
  ans <- !length(new_note)
  if (verbose && !ans) {
    new_note_body <- unlist(lapply(names(new_note), function(check) {
      c(paste("* checking", check), new_note[[check]])
    }))
    message(sprintf("There %s %s unexpected NOTE%s in R CMD check%s\n\n%s",
                    if(length(new_note)>1L) "are" else "is",
                    length(new_note),
                    if(length(new_note)>1L) "s" else "",
                    paste(c("",new_note_flags), collapse=" "),
                    paste(new_note_body, collapse="\n")))
  }
  ans
}

pkg <- if (length(args <- commandArgs(trailingOnly=TRUE))) args[[1]]
rcheck_dir <- Sys.glob(if (is.null(pkg)) "*.Rcheck" else paste(pkg, ".Rcheck", sep=""))
if (length(rcheck_dir)!=1L) stop("This script should be started from directory in which *.Rcheck directory is located, and there must be single Rcheck directory.")

log_file <- file.path(rcheck_dir, "00check.log")
res <- tools:::check_packages_in_dir_results(logs = log_file)[[1L]]
if (length(res)) {
  if (res$status %in% c("FAIL","ERROR","WARN")) {
    status <- 1
    meta <- tools:::analyze_check_log(log_file)
    # print whole file
    message(sprintf("R CMD check%s completed with %s\n\n%s",
                    paste(c("",meta[["Flags"]]), collapse=" "),
                    res$status,
                    paste(readLines(log_file, warn=FALSE), collapse="\n")))
  } else if (res$status=="NOTE") {
    details <- tools:::check_packages_in_dir_details(logs = log_file)
    # print notes which are not whitelisted
    note_ok <- check_note(details, whitelist)
    status <- if (note_ok) 0 else 1
  } else if (res$status=="OK") {
    status <- 0
  } else stop("Unknown status from R CMD check: ", toString(res$status))
} else stop("Unknown results from R CMD check")

if (status==0) cat(sprintf("INFO: %s package R CMD check output successfully validated.\n", sub("\\.Rcheck$", "", rcheck_dir)))

if (!interactive()) q("no", status=status)

File: scripts/validate_r_cmd_check_output.py
#!/usr/bin/python

#
# CRAN submissions use the R CMD CHECK --as-cran approach.
# But that unfortunately does not do a good job of flagging errors in a way that can be automated.
#
# This tool goes combs through the output and returns 0 if it's good and nonzero if it's bad.
#

import sys
import os
import re


class Check:
    def __init__(self, file_name):
        self.file_name = file_name
        self.lineno = 0

    def parse_error(self, message, s, f):
        print("ERROR " + message + " " + self.file_name + " line " + str(self.lineno))
        sys.stdout.write("    >>> " + s)
        s = f.readline()
        while (len(s) > 0):
            sys.stdout.write("    >>> " + s)
            s = f.readline()
        sys.exit(1)

    def process(self):
        # print("Processing " + self.file_name + "...")

        f = open(self.file_name, "r")

        allowed_regex_list = [
            r"^\* using log directory",
            r"^\* using R version",
            r"^\* using R Under development",
            r"^\* R was compiled by",
            r"^    Apple clang.*",
            r"^    gcc.*",
            r"^    GNU Fortran.*",
            r"^\* using platform",
            r"^\* using platform:.*",
            r"^\* running under:.*",
            r"^\* using session charset",
            r"^\* using option .*",
            r"^\* checking .* \.\.\. OK",
            r"^\* checking .* \.\.\. \[\d+s/\d+s\] OK",

            r"^\* checking extension type ... Package",
            r"^\* this is package",
            r"^\* checking CRAN incoming feasibility \.\.\.",
            r"^\*\* found \\donttest examples: .*",
            r"^Maintainer:",
            r"^New maintainer:",
            r"^\s*The H2O.ai team .*",
            r"^Version contains large components .*",
            r"^Insufficient package version .*",
            r"^\Days since last update: .*",
            r"^Old maintainer\(s\):",
            r"^\s*Tom Kraljevic .*",
            r"^NOTE: There was 1 note.",

            r"^\* checking DESCRIPTION meta-information \.\.\.",
            r"^Author field differs from that derived from Authors@R",
            r"^\s*Author: .*",
            r"^\s*Authors@R: .*",

            r"^\n",
            r"^New submission",

            r"^Package was archived on CRAN",
            r"^CRAN repository db overrides:",
            r"^  X-CRAN-Comment: Archived on 2014-09-23 as did not comply with CRAN",
            r"^    policies on use of multiple threads.",

            r"^\* checking installed package size ... NOTE",
            r"^  installed size is .*Mb",
            r"^  sub-directories of 1Mb or more:",
            r"^    java  .*Mb",
            r"^    R  .*Mb",
            r"^    help  .*Mb",
            r"^NOTE: There were 2 notes.",
            #r"^Status: 1 WARNING, 1 NOTE",
            #r"^Status: 1 WARNING, .* NOTEs",
            r"^Status: 2 NOTEs",
            r"^Status: 1 NOTE",
            r"^See",
            r"^ .*/h2o-r/h2o\.Rcheck/00check\.log.*",
            r"^for details.",
            r"^  Running .*",
            r"^ OK",

            r"^Package has FOSS license, installs .class/.jar but has no 'java' directory.",
            r"^Size of tarball: .* bytes",
            r"^\* DONE",

            r"^The Date field is over a month old.*",
            r"^Checking URLs requires 'libcurl' support in the R build",
            # relation to optional data.table package
            r"^\* checking package dependencies ... NOTE",
            r"^Package suggested but not available for checking*",
            r"^\* checking Rd cross-references ... NOTE",
            r"^Package unavailable to check Rd xrefs*",
            r"^Status: 3 NOTEs",
            r"^Status: 4 NOTEs",
            r"^\* checking for future file timestamps ... NOTE",
            r"^unable to verify current time",
            ]

        s = f.readline()
        while (len(s) > 0):
            self.lineno = self.lineno + 1

            allowed = False
            for regex in allowed_regex_list:
                match_groups = re.search(regex, s)
                if (match_groups is not None):
                    # This line is allowed.
                    allowed = True
                    break

            if (not allowed):
                self.parse_error("Illegal output found", s, f)

            s = f.readline()


def main(argv):
    if (not os.path.exists("h2o.Rcheck")):
        print("ERROR:  You must run this script inside the generated R package source directory.")
        sys.exit(1)

    c = Check("h2o.Rcheck/00check.log")
    c.process()

    # No failures.
    sys.exit(0)


if __name__ == "__main__":
    main(sys.argv)
